semantic_conventions/resource/cloud_provider/gcp/gce.yaml

```diff
@@ -0,0 +1,22 @@
+groups:
+  - id: gcp.gce
+    prefix: gcp.gce
+    type: resource
+    brief: >
+      Resources used by Google Compute Engine (GCE).
+    attributes:
+      - id: instance.name
```

Comment:
I don't have experience with Google Cloud but I tried following and reading all the discussions in linked issues/PRs, so I could review it. 

If I got it right, the summary of what we have today is:

- We [populate the OTel](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bc239b5b1f7f9837700855de8e8a5d1d9104df03/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L71) `host.name` attribute with the `InstanceName` in both gce and gke


But, there's also the chance/desire of using the GCE/GKE host name. That is because one could define a [custom one](https://github.com/open-telemetry/opentelemetry-specification/pull/3384#issuecomment-1505747706) and it would be valuable to have it?

So, with that in mind, looking at the PR, I'm a bit confused:

- Why `instance.name` is populated with the host name coming from GCE (the internal DNS or a custom one defined by the user)? 
- Why are we introducing another host name under `gcp.gce.instance.hostname`?

If I understood the discussions and the motive of the change, wouldn't it make more sense to:

- Set `host.name` in OTel with the value from GCE internal DNS/custom one?
- Introduce a new attribute to capture the instance, as you have today?

Are the current changes because of the way `GCEHostName` returns the `InstanceName` and not the HostName? Are those processors/instrumentation stable that we can't break them? Couldn't we change it to record HostName instead of InstanceName?

I think going ahead with the PR as is is very confusing. Won't we essentially end up with 2 " host name" attributes: 

1. The one in the collector https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bc239b5b1f7f9837700855de8e8a5d1d9104df03/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L71
2. Now `gcp.gce.instance.hostname`

Sorry if I'm just being stupid and didn't get things 🙃

Reply:
Thanks for the info! I totally get what you say. But a counter argument can still be made that users can also put whatever under the new hostname attribute right? My concern is that we have the generic attribute, and then each vendor/technology goes and duplicate it, becoming also a "mess".

My point was more of, should we use `host.name` but in the gcp semconv add guidelines to what should be put there, so everyone can benefit from a consistent value? Or my assumption that `host.name` and `instance.name` are completely different things? My understanding is that today, in the collector we set `host.name` to `instance.name` so I'm a bit confused why we need it twice.

Reply:
> But a counter argument can still be made that users can also put whatever under the new hostname attribute right?

If we are precise enough then I don't think that would be a concern, if someone misuses `gcp.instance.hostname` I can point at this repo and tell them "what you did goes against the spec", while with `host.name` there's not much I can do (I think you can argue any of the values on the list above is a valid value for `host.name` according the spec).

> My concern is that we have the generic attribute, and then each vendor/technology goes and duplicate it, becoming also a "mess"

If I were to re-do things I would either drop `host.name` or restrict it to have one of the individual meanings (probably just "a custom value set by the user"). I feel like that ship has sailed though (at least for the first one), so adding new more precise values seems like the best course.

> My point was more of, should we use host.name but in the gcp semconv add guidelines to what should be put there, so everyone can benefit from a consistent value

That seems like an improvement from the current state, but there would still be ambiguity (a `host.name` value coming from the `hostname` command is valid today in GCP and would remain valid after those guidelines).

> My understanding is that today, in the collector we set host.name to instance.name so I'm a bit confused why we need it twice.

I agree there is some redundancy, but again it seems hard to not have that redundancy if we are to keep backwards compatibility.

Reply:
>That seems like an improvement from the current state, but there would still be ambiguity (a host.name value coming from the hostname command is valid today in GCP and would remain valid after those guidelines).

Exactly, but then in these conventions, we say, hey use `host.name` but for GCP we think you can get the better value by populating it with xyz and not whatever other command. I feel that's possible to re-use the generic attributes but under this Cloud environment give it a different purpose. 

Then it would be the same reasoning you mentioned:

> if someone misuses gcp.instance.hostname I can point at this repo and tell them "what you did goes against the spec"

If someone misuse `host.name` by not following the conventions here, then we can say hey you will get more value by setting it this way.

Reply:
Sorry I missed this yesterday. I think a lot of the points have been covered, but to answer a few of the questions:

* the resource detectors are stable, so changing them would be breaking at this point
* the `host.name` attribute is vaguely defined enough that using instance name fits. this is also what most GCP users will expect when they think of the "name of the host". 
* It was [suggested in the other PR](https://github.com/open-telemetry/opentelemetry-specification/pull/3384#issuecomment-1538377030) that we could update the definition or consider adding at `host.dns_name`, but if that happens we can backfill it with these vendor values instead of blocking on that
* these attributes are intended to provide the values explicitly as they are defined within GCP, not to imply that these definitions are any "better" than what is available

it's unfortunate imo that the attribute "host.name" is overloaded with the canonical term "hostname", but as they are defined these are essentially just homonyms. these attributes make a clearer distinction without breaking existing users

Reply:
Alright. Is it possible to get a real-world example of how these 3 attributes would look like when populated according to this PR? I think everyone would benefit from seeing it "in action".

Reply:
Sure, given the 2 sample VMs below (one with the default hostname and another with a [custom hostname](https://cloud.google.com/compute/docs/instances/custom-hostname-vm)):

![Screenshot 2023-05-18 at 12 36 03 PM](https://github.com/open-telemetry/semantic-conventions/assets/1839101/7a59e3a4-84ee-46dd-8a32-3acfa7e1e097)

The output of `hostname -f` and `hostname` on **instance-1** are:
![Screenshot 2023-05-18 at 12 35 21 PM](https://github.com/open-telemetry/semantic-conventions/assets/1839101/d9c5b450-2475-4e54-9247-b46c3df5f43c)
(gcp project id omitted)

The same output on **instance-2** (showing custom hostname):
![Screenshot 2023-05-18 at 12 36 42 PM](https://github.com/open-telemetry/semantic-conventions/assets/1839101/fa5c7fe6-371c-4c09-9ed5-46f63fa47367)

So for these 2 instances, the attributes would be:
|attr|instance-1|instance-2|
|---|---|---|
|`host.name`|`instance-1`|`instance-2-custom-hostname`|
|`gcp.gce.instance.name`|`instance-1`|`instance-2-custom-hostname`|
|`gcp.gce.instance.hostname`|`instance-1.us-central1-a.c.<gcp-project>.internal`|`foo-bar.example.com`|

Reply:
Thanks for the examples, really helps painting the picture. 

So, looking now at the example of `gcp.gce.instance.hostname` and its value `instance-1.us-central1-a.c.<gcp-project>.internal` and reading the current recommendations we have for [host.name](https://github.com/open-telemetry/semantic-conventions/blob/0887921aed44ea6fb818e9dc026b4d8ff8cabe5f/semantic_conventions/resource/host.yaml#L19) doesn't it fit well? (emphasis mine)

> Name of the host. On Unix systems, it may contain what the hostname
> command returns, **or the fully qualified hostname**, or another name
> specified by the user.

But I understand that it's not ideal to change now, since the resource detector in the collector reports `gcp.gce.instance.name` for `host.name` and changing that is not good, correct?

So, the "only" benefit here is that you have "GCP-specific" fields to capture "GCP-specific" concepts (instane.name/instance.hostname), even though one (`instance.name`) is already captured on `host.name` WHEN using the resource detector in the collector https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bc239b5b1f7f9837700855de8e8a5d1d9104df03/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L71.

I think if we didn't have the current behavior above today, I feel reporting `gcp.gce.instance.hostname` on `host.name` is still correct. The concern that is breaking I have to say I don't fully understand. The point also about "cloud provider-related dependencies to core SDK libraries", if people run their code on GCP and want to get the "right telemetry experience" they would then install GCP resource detector either directly on their apps (via package) or use the collector. I don't think having the conventions for GCP indicating what to put on `host.name` would require OTel SDKs to do GCP stuff. That is the job of instrumentations. If they run their apps on GCP and don't use resource detectors then well, they will get what they get and I think that's fair. 

> the resource detectors are stable, so changing them would be breaking at this point

Hum, nothing I guess can be done, but the resource semconv where `host.name` are, is still not marked as stable, so maybe marking the resource detector also as stable was a mistake. `Experimental` in OTel allows breaking changes  https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md

I don't want to bikeshed or stall the work you all are putting into this, specially since I'm not an expert. But I'm a bit worried about the duplication and possible confusion this will bring. Let's see what the other folks think about it, maybe it's just be being overly cautious.

Reply:
I think what would be great is if we could achieve this:

|attr|instance-1|instance-2|
|---|---|---|
|`host.name`|`instance-1.us-central1-a.c.<gcp-project>.internal`|`foo-bar.example.com`|
|`gcp.gce.instance.name`|`instance-1`|`instance-2-custom-hostname`|

Then we only need to introduce `gcp.gce.instance.name`. 

Also: I just looked at https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/bc239b5b1f7f9837700855de8e8a5d1d9104df03/processor/resourcedetectionprocessor and it says stability is `beta`. Does the gcp component internally has different stability level? I couldn't find it 🤔

Reply:
The sdk libraries for GCP resource detection are what is stable: https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/tree/detectors/gcp/v1.13.1 

(https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/ceb4457c9caae3705d3f9f85b8fbf6c4c0e6f0e7/tools/release.go#L32)

In hindsight, using the FQDN for `host.name` would have worked, but using the instance name also works within the definition of the attribute. Since *most* users care about the instance name, I think that was the most practical and applicable choice at the time.

Reply:
> Is your blocking request really going to improve o11y materially?

Probably not, it will be data duplication which will cost money but maybe that's not a big problem given it's a resource attribute. If no one see this as a problem them I'm also fine with it. 

> As a compromise, would adding just gcp.gce.instnace.hostname make sense here? This way we don't have duplicated attributes.

If I got it right, the intent is to have a specific attribute for instance because it's not clear what `host.name` has so I guess the desire of having well defined gcp attributes to define that. 

I will remove my blocking review now that I understand it more, so work can move forward.

Reply:
Thanks @joaopgrassi.  Regarding duplication, that's what my compromise here is about.  We would get the full-domain instance name and the short instance name that exists today for users. 

We could also add language for `gcp.gce.instance.hostname` to only be provided when it's different from `host.name`.